accepted/system-range/system-range.md

```diff
@@ -0,0 +1,271 @@
+{
+    public string this[Range range] => Substring(range.Index, range.Length);
+}
+```
+
+This allows creating substrings simply by using the indexer:
+
+```csharp
+var helloWorld = "Hello, World!";
+var hello = helloWorld[..5];
+var world = helloWorld[7..12];
+```
+
+While this might appear to just be minor syntactic sugar it allows for creating
+powerful APIs, such as the indexer for `Tensor<T>` (which you can think of as
+multidimensional array where the dimensions aren't statically known).
+
+```csharp
+partial class Tensor<T>
+{
+    public Tensor<T> this[params Range[] ranges] { get; }
+}
+```
+
+The indexer takes a list of ranges that can be used to get slices of the tensor.
+Imagine you have a two dimensional tensor (in other words a table). Imagine it
+represents test data where each row is a test case and the columns are
+representing the inputs with the last column containing the expected result. Now
+let's say you want to split this data into two tensors, one representing the
+inputs and one the outputs.
+
+This might look like this:
+
+```csharp
+Tensor<float> testDataTable = GetTestData();
+int columnCount = testDataTable.GetLength(1);
+int outputColumn = columnCount - 1;
+
+Tensor<float> inputTable = testDataTable[.., ..outputColumn];
+Tensor<float> outputTable = testDataTable[.., outputColumn..];
+```
+
+## Requirements
+
+### Goals
+
+* We need a type that both, plays well with existing concepts in the base class
+  library, as well as the intended syntax for C#
+* We need to ensure we can extend the notion of ranges in the base class library
+  in the future (for example, by supporting negative ranges or element types
+  other than `Int32`).
+
+### Non-Goals
+
+* Supporting other numeric types or even offer a generic range type
+* Creating overloads on existing types that will accept `Range` (that doesn't
+  mean we shouldn't be doing it, but that it will be covered by a separate
+  design proposal).
+
+## Design
+
+### Proposed API
+
+```csharp
+namespace System {
+    public struct Range : IEnumerable, IEnumerable<int>, IEquatable<Range> {
+        public static Range Empty { get; }
+        public static Range All { get; }
+        public static Range Between(int start, int end);
+        public static Range From(int start);
+        public static Range To(int end);
+        public Range(int index, int length);
+        public int Index { get; }
+        public int Length { get; }
+        public bool IsClosed { get; }
+        public Range Close(Range range);
+        public override string ToString();
+        public override int GetHashCode();
+        public bool Equals(Range other);
+        public override bool Equals(object other);
+        public static bool operator==(Range left, Range right);
+        public static bool operator!=(Range left, Range right);
+    }
+}
+```
+
+### Companion APIs
+
+```csharp
+namespace System {
+    public partial class String {
+        public string this[Range range] { get; }
```

Comment:
Why does this return a string vs. a `Span<char>`?

Reply:
Because the vast majority of our customers will not care for getting spans out of `string` but would really enjoy the shorthand for `Substring`. For the folks that want spans, they use `str.AsSpan(start:end)`.